diff --git a/hybrid/hybrid_quick.py b/hybrid/hybrid_quick.py
index c888328..40d02a3 100644
--- a/hybrid/hybrid_quick.py
+++ b/hybrid/hybrid_quick.py
@@ -398,7 +398,9 @@ def sub_bigq_quick_u_q(n,k,h,u,q):
                 log_G3 = 1000
                 G2 = 2 ** 1000
                 G3 = 2 ** 1000
-            tot = -log_pr + log2(G1 + G2 + G3 * (2 ** try_comp))
+            log_terms = (log_G1, log2(G2), log2(G3) + try_comp)
+            log_max = max(log_terms)
+            tot = -log_pr + log_max + log2(sum(2 ** (x - log_max) for x in log_terms))
         if tot < best:
             best = tot
